tools/refguide_check.py

```diff
@@ -812,39 +814,33 @@ def flush(self):
                 sys.stdout.flush()
 
     # Run tests, trying to restore global state afterward
-    old_printoptions = np.get_printoptions()
-    old_errstate = np.seterr()
-    old_stderr = sys.stderr
-    cwd = os.getcwd()
-    tmpdir = tempfile.mkdtemp()
-    sys.stderr = MyStderr()
-    try:
-        os.chdir(tmpdir)
-
-        # try to ensure random seed is NOT reproducible
-        np.random.seed(None)
-
-        ns = {}
-        for t in tests:
-            # We broke the tests up into chunks to try to avoid PSEUDOCODE
-            # This has the unfortunate side effect of restarting the global
-            # namespace for each test chunk, so variables will be "lost" after
-            # a chunk. Chain the globals to avoid this
-            t.globs.update(ns)
-            t.filename = short_path(t.filename, cwd)
-            # Process our options
-            if any([SKIPBLOCK in ex.options for ex in t.examples]):
-                continue
-            fails, successes = runner.run(t, out=out, clear_globs=False)
-            if fails > 0:
-                success = False
-            ns = t.globs
-    finally:
-        sys.stderr = old_stderr
-        os.chdir(cwd)
-        shutil.rmtree(tmpdir)
-        np.set_printoptions(**old_printoptions)
-        np.seterr(**old_errstate)
+    with np.errstate(), np.printoptions(), redirect_stderr(MyStderr()):
```

Comment:
`redirect_stderr` is available in python 3.5.
sadly there isn't a `temp_cwd` contextmanager yet

Reply:
Probably worth making your own context manager for that in this file - it'll reduce the nesting.

Reply:
I'm happy to add it to my longer term list but I'd prefer to not do it today/this week.

Reply:
Ok if you don't want to do it, but with `contextlib.contextmanager` all it requires is moving this try/except to it's own function, replacing it's contents with just `yield`, and adding to the `with`.

Reply:
Okay, you convinced me. I'll give it a shot.